drivers/setup/blue/blue.c

```diff
@@ -1113,112 +1318,201 @@ ScrIoControl(
+            if (pCoord == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                break;
+            }
+            /* Capture the input info data */
+            dwCoord = *pCoord;
+
+            nMaxLength = stk->Parameters.DeviceIoControl.OutputBufferLength - sizeof(COORD);
+            Irp->IoStatus.Information = 0;
+
+            if ( dwCoord.X < 0 || dwCoord.X >= DeviceExtension->Columns ||
+                 dwCoord.Y < 0 || dwCoord.Y >= DeviceExtension->Rows    ||
+                 nMaxLength == 0 )
             {
-                Irp->IoStatus.Information = 0;
                 Status = STATUS_SUCCESS;
                 break;
             }
 
+            pChar = (PCHAR)(pCoord + 1);
+
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
                 vidmem = DeviceExtension->VideoMemory;
-                offset = (pCoord->X + pCoord->Y * DeviceExtension->Columns) * 2;
+                offset = (dwCoord.X + dwCoord.Y * DeviceExtension->Columns) * 2;
 
-                nMaxLength = min(stk->Parameters.DeviceIoControl.OutputBufferLength - sizeof(COORD),
-                                 (DeviceExtension->Rows - pCoord->Y)
-                                    * DeviceExtension->Columns - pCoord->X);
+                nMaxLength = min(nMaxLength,
+                                 (DeviceExtension->Rows - dwCoord.Y)
+                                    * DeviceExtension->Columns - dwCoord.X);
 
                 for (dwCount = 0; dwCount < nMaxLength; dwCount++, pChar++)
                 {
                     vidmem[offset + (dwCount * 2)] = *pChar;
                 }
+                Irp->IoStatus.Information = dwCount * sizeof(CHAR);
             }
 
-            Irp->IoStatus.Information = 0;
             Status = STATUS_SUCCESS;
             break;
         }
 
         case IOCTL_CONSOLE_DRAW:
         {
-            PCONSOLE_DRAW ConsoleDraw;
+            CONSOLE_DRAW ConsoleDraw;
+            PCONSOLE_DRAW pConsoleDraw;
             PUCHAR Src, Dest;
             UINT32 SrcDelta, DestDelta, i;
 
+            /* Validate output buffer */
+            if (stk->Parameters.DeviceIoControl.OutputBufferLength < sizeof(CONSOLE_DRAW))
+            {
+                Status = STATUS_INVALID_PARAMETER;
+                break;
+            }
+            ASSERT(Irp->MdlAddress);
+            pConsoleDraw = MmGetSystemAddressForMdlSafe(Irp->MdlAddress, NormalPagePriority);
+            if (pConsoleDraw == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                break;
+            }
+            /* Capture the input info data */
+            ConsoleDraw = *pConsoleDraw;
+
+            /* Check whether we have the size for the header plus the data area */
+            if (stk->Parameters.DeviceIoControl.OutputBufferLength - sizeof(CONSOLE_DRAW)
+                    < ConsoleDraw.SizeX * ConsoleDraw.SizeY * 2)
```

Comment:
What I did here is threefold:
- First, since all the console screenbuffer sizes and positions are stored and provided in USHORTs (so, minimum value 0, maximum value 0xFFFF), I converted the CONSOLE_DRAW structure to use USHORTs instead of ULONGs.
- Second, since 0xFFFF * 0xFFFF = 0xFFFE0001 (when the ushorts have been zero-extended to ulongs) this also holds without overflowing into a ULONG result.
- Third because the `OutputBufferLength - sizeof(....)` is usually even I instead do a division by 2, so that we get: `(OutputBufferLength - sizeof(....)) / 2 < SizeX * SizeY`

Please tell me what you think about this.

Reply:
Looks good.